extensions/amp-gwd-animation/0.1/amp-gwd-animation.js

```diff
@@ -0,0 +1,228 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {ActionTrust} from '../../../src/action-trust';
+import {CSS} from '../../../build/amp-gwd-animation-0.1.css';
+import {Services} from '../../../src/services';
+import {getServiceForDoc} from '../../../src/service';
+import {
+  GWD_SERVICE_NAME,
+  GWD_TIMELINE_EVENT,
+  installGwdRuntimeServiceForDoc,
+} from './amp-gwd-animation-impl';
+import {isExperimentOn} from '../../../src/experiments';
+import {user} from '../../../src/log';
+
+/** @const {string} The custom element tag name for this extension. */
+export const TAG = 'amp-gwd-animation';
+
+/** @const {string} Experiment flag name. TODO(sklobovskaya): Remove. */
+export const EXPERIMENT = TAG;
+
+/** @const {string} */
+export const GWD_PAGEDECK_ID = 'pagedeck';
+
+/**
+ * Actions supported by the extension and the parameters each requires, as a
+ * string identifying the path to the data in an AMP action invocation object.
+ * For example, 'args.id' will map to the data at `invocation.args.id`.
+ * Each action name currently corresponds to an identically-named method in the
+ * GWD runtime service, which is invoked with the evaluated arguments
+ * (@see getActionImplArgs and createAction_).
+ * @const {!Object<string, !Array<string>>}
+ */
+const ACTION_IMPL_ARGS = {
+  'play': ['args.id'],
+  'pause': ['args.id'],
+  'togglePlay': ['args.id'],
+  'gotoAndPlay': ['args.id', 'args.label'],
+  'gotoAndPause': ['args.id', 'args.label'],
+  'gotoAndPlayNTimes': ['args.id', 'args.label', 'args.N', 'event.eventName'],
+  'setCurrentPage': ['args.index'],
+};
+
+/**
+ * Given an action name, extracts its required arguments from an invocation
+ * payload, which may contain the data needed by this action in its `args` or
+ * `events` child objects. The arguments required by each action are specified
+ * in ACTION_IMPL_ARGS as path strings.
+ * @param {string} actionName
+ * @param {!../../../src/service/action-impl.ActionInvocation} invocation
+ * @return {!Array<string>}
+ */
+const getActionImplArgs = function(actionName, invocation) {
+  const argDefs = user().assert(
+      ACTION_IMPL_ARGS[actionName],
+      `The action ${actionName} is not a supported action.`);
+
+  return argDefs.map(argDef => {
+    // Walk the invocation object to get the requested property by
+    // its path, e.g., 'args.id' == invocation.args.id
+    let obj = invocation;
+    const props = argDef.split('.');
```

Comment:
Please see if this utility does the same for you: https://github.com/ampproject/amphtml/blob/master/src/json.js#L82

Reply:
Yes, that works great. Thanks!